crates/driver/src/domain/competition/sorting.rs

```diff
@@ -0,0 +1,122 @@
+use {
+    crate::{
+        domain::{
+            competition::{auction::Tokens, order},
+            eth,
+        },
+        util,
+    },
+    chrono::{Duration, Utc},
+    std::{cmp::Ordering, sync::Arc},
+};
+
+pub trait OrderComparator: Send + Sync {
+    fn compare(
+        &self,
+        order_a: &order::Order,
+        order_b: &order::Order,
+        tokens: &Tokens,
+        solver: &eth::H160,
+    ) -> Ordering;
+}
+
+impl<F> OrderComparator for F
+where
+    F: Fn(&order::Order, &order::Order, &Tokens, &eth::H160) -> Ordering + Send + Sync,
+{
+    fn compare(
+        &self,
+        order_a: &order::Order,
+        order_b: &order::Order,
+        tokens: &Tokens,
+        solver: &eth::H160,
+    ) -> Ordering {
+        self(order_a, order_b, tokens, solver)
+    }
+}
+
+pub trait OrderingKey: Send + Sync + 'static {
+    type Key: Ord + Send + Sync + 'static;
+
+    fn key(&self, order: &order::Order, tokens: &Tokens, solver: &eth::H160) -> Self::Key;
+
+    /// Returns a comparator that compares two orders based on the key in
+    /// reverse order.
+    fn into_comparator(self) -> Arc<dyn OrderComparator>
+    where
+        Self: Sized,
+    {
+        Arc::new(
+            move |a: &order::Order, b: &order::Order, tokens: &Tokens, solver: &eth::H160| {
+                self.key(a, tokens, solver)
+                    .cmp(&self.key(b, tokens, solver))
+                    .reverse()
+            },
+        )
+    }
+}
+
+pub struct OrderClass;
+impl OrderingKey for OrderClass {
+    type Key = i32;
+
+    // Market orders are preferred over limit orders, as the expectation is that
+    // they should be immediately fulfillable. Liquidity orders come last, as they
+    // are the most niche and rarely used.
+    fn key(&self, order: &order::Order, _tokens: &Tokens, _solver: &eth::H160) -> Self::Key {
+        match order.kind {
+            order::Kind::Market => 2,
+            order::Kind::Limit { .. } => 1,
+            order::Kind::Liquidity => 0,
+        }
+    }
+}
+
+pub struct ExternalPrice;
+impl OrderingKey for ExternalPrice {
+    type Key = num::BigRational;
+
+    fn key(&self, order: &order::Order, tokens: &Tokens, _solver: &eth::H160) -> Self::Key {
+        order.likelihood(tokens)
+    }
+}
+
+pub struct CreationTimestamp {
+    pub threshold: Duration,
+}
+impl OrderingKey for CreationTimestamp {
+    type Key = Option<util::Timestamp>;
+
+    fn key(&self, order: &order::Order, _tokens: &Tokens, _solver: &eth::H160) -> Self::Key {
+        (order.created.0
+            > u32::try_from((Utc::now() - self.threshold).timestamp()).unwrap_or(u32::MAX))
```

Comment:
Nit: storing this as `earliest_allowed_creation` or so would help with the readability.